worldedit-core/src/main/java/com/boydti/fawe/beta/implementation/lighting/HeightMapType.java

```diff
@@ -1,5 +1,46 @@
 package com.boydti.fawe.beta.implementation.lighting;
 
+import com.sk89q.worldedit.registry.state.Property;
+import com.sk89q.worldedit.registry.state.PropertyKey;
+import com.sk89q.worldedit.world.block.BlockCategories;
+import com.sk89q.worldedit.world.block.BlockState;
+
 public enum HeightMapType {
-    MOTION_BLOCKING, MOTION_BLOCKING_NO_LEAVES, OCEAN_FLOOR, WORLD_SURFACE
+    MOTION_BLOCKING {
+        @Override
+        public boolean blocks(BlockState state) {
+            return state.getMaterial().isSolid() || HeightMapType.isFluid(state);
+        }
+    },
+    MOTION_BLOCKING_NO_LEAVES {
+        @Override
+        public boolean blocks(BlockState state) {
+            return (state.getMaterial().isSolid() || HeightMapType.isFluid(state)) && !HeightMapType.isLeaf(state);
+        }
+    },
+    OCEAN_FLOOR {
+        @Override
+        public boolean blocks(BlockState state) {
+            return state.getMaterial().isSolid();
+        }
+    },
+    WORLD_SURFACE {
+        @Override
+        public boolean blocks(BlockState state) {
+            return !state.isAir();
+        }
+    };
+
+    private static boolean isLeaf(BlockState state) {
+        return BlockCategories.LEAVES.contains(state);
+    }
+
+    private static boolean isFluid(BlockState state) {
```

Comment:
isFluidOrWaterlogged or similar might be better. Could also be worth having this as a public utility-method elsewhere for various commands (in particular I know Matt's looking at underwater //flora)

Reply:
I tried to keep it similar to NMS getFluid().isEmpty(). Might be better to call it hasFluid or isFluidOrWaterlogged, but hasFluid might be better (?)

Reply:
hasFluid sounds better yeah. Adding some javadoc annotation stating explicitly what is included would be good I think as well.